ledger/src/ui/test/signer_authorization/test_signer_authorization.c

```diff
@@ -0,0 +1,449 @@
+    "\x8f\xda\x83\xb6\xe2\x1d\xd3\xac\x65\x56\xe0\x53\x8f\x2e\xa8\x05\x4a\x8c"
+    "\x43\x26\x8a",
+
+    "\xf9\x6f\x6b\x3e\x3c\xe5\x08\x8b\x46\x83\xbe\x87\x3c\x07\x5a\x61\x4a\x8c"
+    "\x65\x64\x9a\x4c\xa8\x6a\x78\xa0\xc6\x6c\x9e\x47\x7e\x20\x5f\x32\x2e\xc9"
+    "\x89\x9e\x0f\x4d\x00\x60\xab\x30\x69\x28\xcd\x69\x43\x01\x30\x36\x01\x33"
+    "\xf1\x32\x26\xcd\x1c\x62\x3b\x12\xf7\xf4\xef\x69\x87\x8b\xe8\x7d\xbe\x60"
+    "\x86\x4d\xa2",
+
+    "\x4b\xf1\xc6\x36\xef\x35\x06\x39\x42\x9e\x3e\xf4\x02\xd3\xd1\x63\xa0\x59"
+    "\x73\x8d\x39\x3d\xd9\x38\xa0\x34\xa3\x2e\xf7\xf8\xe2\xbe\xe7\x2d\xbd\x07"
+    "\xf0\x8c\x00\xc0\x5c\x62\xee\x1a\x5a\xf4\x24\xe1\x9d\x03\x83\x36\x1b\xbf"
+    "\x3e\x7f\x55\x61\xf0\x52\x42\x21\x1d\xea\xe8\xfc\x26\x5d\xa1\x83\x04\xc7"
+    "\xe0\xca\x4e",
+};
+
+static const uint8_t G_authorizers_pubkeys[][AUTHORIZED_SIGNER_PUBKEY_LENGTH] =
+    AUTHORIZERS_PUBKEYS;
+static const int G_num_pubkeys =
+    sizeof(G_authorizers_pubkeys) / sizeof(G_authorizers_pubkeys[0]);
+
+static const unsigned char G_authorized_signer_hash[] =
+    "AUTHORIZED HASH 123456789abcdef";
+
+// Helper functions
+static bool is_key_authorized(cx_ecfp_public_key_t *key) {
+    for (int i = 0; i < G_num_pubkeys; i++) {
+        if (0 == memcmp(G_authorizers_pubkeys[i], key->W, PUBKEYSIZE)) {
+            return true;
+        }
+    }
+    return false;
+}
+
+static unsigned char *get_signature(cx_ecfp_public_key_t *key) {
+    for (int i = 0; i < G_num_pubkeys; i++) {
+        if (0 == memcmp(G_authorizers_pubkeys[i], key->W, PUBKEYSIZE)) {
+            return (unsigned char *)G_valid_signatures[i];
+        }
+    }
+    ASSERT_FAIL();
+}
+
+// Mock functions
+void os_memmove(void *dst, const void *src, unsigned int length) {
+    memmove(dst, src, length);
+}
+
+int cx_keccak_init(cx_sha3_t *hash, int size) {
+    // Size is passed in bits for this call
+    int size_in_bytes = size / 8;
+    assert(size_in_bytes <= sizeof(hash->hash));
+
+    hash->size_in_bytes = size_in_bytes;
+    memset(hash->hash, 0xff, sizeof(hash->hash));
+    return 0;
+}
+
+int cx_hash(cx_hash_t *hash,
+            int mode,
+            unsigned char *in,
+            unsigned int len,
+            unsigned char *out) {
+    // We don't need an actual hash algorithm for the purpose of this test,
+    // so we just XOR the input with the current hash to make sure the correct
+    // data was passed
+    assert(len <= hash->size_in_bytes);
+    for (int i = 0; i < len; i++) {
+        hash->hash[i] ^= in[i];
+    }
+
+    if (mode & CX_LAST) {
+        memcpy(out, hash->hash, hash->size_in_bytes);
+    }
+    return 0;
+}
+
+int cx_ecfp_init_public_key(cx_curve_t curve,
+                            unsigned char *rawkey,
+                            unsigned int key_len,
+                            cx_ecfp_public_key_t *key) {
+    assert(NULL != rawkey);
+    assert(NULL != key);
+    assert(CX_CURVE_256K1 == curve);
+    assert(65 == key_len);
```

Comment:
Isn't the raw key here meant to be the private key? In that case, it would be 32 bytes long.

Reply:
Actually, this is the public key. All this function does is initialize a `cx_ecfp_public_key_t` struct with the given raw key. This function is called on` signer_authorization.c` as:

`cx_ecfp_init_public_key(CX_CURVE_256K1, (unsigned char*)authorizers_pubkeys[i], sizeof(authorizers_pubkeys[i]), &sigaut_ctx->pubkey);`

Reply:
Cool, thanks for clarifying.